// n8n-workflow/dist/Extensions/NumberExtensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.numberExtensions = void 0;
const ExpressionError_1 = require("./../ExpressionError");
function format(value, extraArgs) {
    const [locales = 'en-US', config = {}] = extraArgs;
    return new Intl.NumberFormat(locales, config).format(value);
}
function isEven(value) {
    if (!Number.isInteger(value)) {
        throw new ExpressionError_1.ExpressionExtensionError('isEven() is only callable on integers');
    }
    return value % 2 === 0;
}
function isOdd(value) {
    if (!Number.isInteger(value)) {
        throw new ExpressionError_1.ExpressionExtensionError('isOdd() is only callable on integers');
    }
    return Math.abs(value) % 2 === 1;
}
function floor(value) {
    return Math.floor(value);
}
function ceil(value) {
    return Math.ceil(value);
}
function round(value, extraArgs) {
    const [decimalPlaces = 0] = extraArgs;
    return +value.toFixed(decimalPlaces);
}
ceil.doc = {
    name: 'ceil',
    description: 'Rounds up a number to a whole number',
    returnType: 'number',
};
floor.doc = {
    name: 'floor',
    description: 'Rounds down a number to a whole number',
    returnType: 'number',
};
isEven.doc = {
    name: 'isEven',
    description: 'Returns true if the number is even. Only works on whole numbers.',
    returnType: 'boolean',
};
isOdd.doc = {
    name: 'isOdd',
    description: 'Returns true if the number is odd. Only works on whole numbers.',
    returnType: 'boolean',
};
format.doc = {
    name: 'format',
    returnType: 'string',
};
round.doc = {
    name: 'round',
    returnType: 'number',
};
exports.numberExtensions = {
    typeName: 'Number',
    functions: {
        ceil,
        floor,
        format,
        round,
        isEven,
        isOdd,
    },
};
