// n8n-workflow/dist/Extensions/ObjectExtensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.objectExtensions = exports.urlEncode = exports.compact = void 0;
const ExpressionError_1 = require("../ExpressionError");
function isEmpty(value) {
    return Object.keys(value).length === 0;
}
function isNotEmpty(value) {
    return !isEmpty(value);
}
function hasField(value, extraArgs) {
    const [name] = extraArgs;
    return name in value;
}
function removeField(value, extraArgs) {
    const [name] = extraArgs;
    if (name in value) {
        const newObject = { ...value };
        delete newObject[name];
        return newObject;
    }
    return value;
}
function removeFieldsContaining(value, extraArgs) {
    const [match] = extraArgs;
    if (typeof match !== 'string' || match === '') {
        throw new ExpressionError_1.ExpressionExtensionError('removeFieldsContaining(): expected non-empty string arg');
    }
    const newObject = { ...value };
    for (const [key, val] of Object.entries(value)) {
        if (typeof val === 'string' && val.includes(match)) {
            delete newObject[key];
        }
    }
    return newObject;
}
function keepFieldsContaining(value, extraArgs) {
    const [match] = extraArgs;
    if (typeof match !== 'string' || match === '') {
        throw new ExpressionError_1.ExpressionExtensionError('argument of keepFieldsContaining must be a non-empty string');
    }
    const newObject = { ...value };
    for (const [key, val] of Object.entries(value)) {
        if (typeof val !== 'string' || (typeof val === 'string' && !val.includes(match))) {
            delete newObject[key];
        }
    }
    return newObject;
}
function compact(value) {
    const newObj = {};
    for (const [key, val] of Object.entries(value)) {
        if (val !== null && val !== undefined && val !== 'nil' && val !== '') {
            if (typeof val === 'object') {
                if (Object.keys(val).length === 0)
                    continue;
                newObj[key] = compact(val);
            }
            else {
                newObj[key] = val;
            }
        }
    }
    return newObj;
}
exports.compact = compact;
function urlEncode(value) {
    return new URLSearchParams(value).toString();
}
exports.urlEncode = urlEncode;
isEmpty.doc = {
    name: 'isEmpty',
    description: 'Checks if the Object has no key-value pairs',
    returnType: 'boolean',
};
isNotEmpty.doc = {
    name: 'isNotEmpty',
    description: 'Checks if the Object has key-value pairs',
    returnType: 'boolean',
};
compact.doc = {
    name: 'compact',
    description: 'Removes empty values from an Object',
    returnType: 'boolean',
};
urlEncode.doc = {
    name: 'urlEncode',
    description: 'Transforms an Object into a URL parameter list. Only top-level keys are supported.',
    returnType: 'string',
};
hasField.doc = {
    name: 'hasField',
    returnType: 'boolean',
};
removeField.doc = {
    name: 'removeField',
    returnType: 'object',
};
removeFieldsContaining.doc = {
    name: 'removeFieldsContaining',
    returnType: 'object',
};
keepFieldsContaining.doc = {
    name: 'keepFieldsContaining',
    returnType: 'object',
};
exports.objectExtensions = {
    typeName: 'Object',
    functions: {
        isEmpty,
        isNotEmpty,
        hasField,
        removeField,
        removeFieldsContaining,
        keepFieldsContaining,
        compact,
        urlEncode,
    },
};
